lambda/shared/src/main/scala/feral/lambda/tracing/TracedLambda.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package feral.lambda.tracing
+
+import cats.data.Kleisli
+import cats.effect.{Trace => _, _}
+import feral.lambda._
+import natchez._
+
+object TracedLambda {
+  def evalKernel[Event] = new PartiallyAppliedEvalKernel[Event]()
+
+  def apply[F[_] : MonadCancelThrow, G[_], Event, Result](entryPoint: EntryPoint[F],
+                                                          extractKernel: Kleisli[F, (Event, Context[F]), Kernel])
+                                                         (lambda: Lambda[G, Event, Result])
+                                                         (implicit LT: LiftTrace[F, G]): Lambda[F, Event, Result] =
+    (event, context) =>
+      Resource
+        .eval(extractKernel((event, context)))
+        .flatMap { kernel => entryPoint.continueOrElseRoot(context.functionName, kernel) }
+        .use(LiftTrace[F, G].run(_)((_: Trace[G]) => lambda(event, context.mapK(LiftTrace[F, G].liftK))))
```

Comment:
@armanbilge were you expecting the `Trace[G]` here to just be discarded? I think it works but it's kind of confusing.

Reply:
Yeah, I'm not crazy about it but I'm still trying to figure out how to deal with `IOLocal`-based tracing. See how I used it in https://github.com/typelevel/feral/blob/7860c4d788161bece6724607ba26e39a033fbfea/lambda-natchez/src/main/scala/feral/lambda/natchez/TracedLambda.scala.